### ISSUE TITLE
Feature: Add ability to stop/interupt stream

### ISSUE THREAD

**Issue by taylorn-ai** [Azure-Samples/azure-search-openai-demo#1221]
### This issue is for a: (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [X] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

I would like to have a button that will stop/interrupt the stream being rendered (ChatGPT has this). So if the response is going downhill, the user would be able to stop it before it goes any further. I've tried implementing something myself, but it doesn't quite work with the ASyncGenerator.

**Comment by pamelafox** [comment]
Interesting. I wonder if you could try ReadableStream.cancel() method, as ReadableStream() is what we use in the frontend:
https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/cancel

The spec says that sending a cancel event should attempt to cancel the underlying mechanism, so I think it would close the TCP connection? I'm not sure if you'd still pay for the full tokens of the call though. That's worth experimenting with, if you haven't yet.

Otherwise I fear you'd need to move to something with better bidirectional support like web sockets.

**Comment by TaylorN15** [comment]
Thanks for the prompt response! I was trying to cancel the asyncio thread in Quart, but cancelling the stream make much more sense :)

I've been able to implement this just now, and it seems to work quite well. As far as I can tell, it also stops the response coming back from the model, but I am not sure if there's a way I can check if I was still billed for the full token count, but it doesn't bother me that much, the user experience is more important :)

```
const [responseStream, setResponseStream] = useState<ReadableStream|null>(null);
const handleAsyncRequest = async (question: string, answers: [string, ChatAppResponse][], setAnswers: Function, responseBody: ReadableStream<any>) => {
        setResponseStream(responseBody);
// ... other code ... //
```

Then I've just added a button that calls onStopClick()

```
const onStopClick = async () => {
    try {
        if (responseStream) {
            responseStream.cancel("User cancelled the stream");
        }
    } catch (e) {
        console.log(e);
    }
};

**Comment by pamelafox** [comment]
Very cool! You might be able to monitor token usage in your Azure OpenAI metrics dashboard.

**Comment by hammadilyes** [comment]
Looks nice, very cool indeed. We will try it on our end and give you feedback!

**Comment by TrentinMarc** [comment]
Hi all, 
@TaylorN15 don't you face an error such as : 

`ReadableStream.cancel: Cannot cancel a stream locked by a reader`  ?

I could not interrupt my chat the way you shared, but I just found a workaround like this : 

``` typescript
export class ClassForReadNDJSONStream {
    reader: ReadableStreamDefaultReader;
    responseBody: ReadableStream<any>;

    constructor(responseBody: ReadableStream<any>) {
        this.responseBody = responseBody;
        this.reader = responseBody.getReader();
    }
    async *readStream() {
        let runningText = "";
        let decoder = new TextDecoder("utf-8");
        while (true) {
            const { done, value } = await this.reader.read();
            if (done) break;
            var text = decoder.decode(value, { stream: true });
            const objects = text.split("\n");
            for (const obj of objects) {
                try {
                    runningText += obj;
                    let result = JSON.parse(runningText);
                    yield result;
                    runningText = "";
                } catch (e) {
                    // Not a valid JSON object
                }
            }
        }
    }
}
```
+ 
```typescript

const c4readNDSJonstream = new ClassForReadNDJSONStream(responseBody);
setResponseStream(c4readNDSJonstream); // Using your method with ClassForReadNDJSONStream signature

for await (const event of c4readNDSJonstream.readStream()) {
// other code ...
```
Based on @pamelafox readNDJSONStream lib but expose reader to allow reader.releaseLock() (and then responseBody.cancel()) in onStopClick() to interrupt chat answer.

What do you think about it ?

**Comment by TaylorN15** [comment]
Yes, actually, I did get that error :)

I ended up implementing an abort controller instead.

**Comment by hammadilyes** [comment]
you might sharing the code ? @TaylorN15

**Comment by TaylorN15** [comment]
Add the AbortController
`const [abortController, setAbortController] = useState<AbortController | null>(null);`

Add the below into `makeApiRequest`
```
const makeApiRequest = async (question: string) => {
	const controller = new AbortController();
	setAbortController(controller);
```

Add the AbortSignal to the props of `handleAsyncRequest`
```
const handleAsyncRequest = async (
	question: string,
	answers: [string, ChatAppResponse][],
	responseBody: ReadableStream<any>,
	signal: AbortSignal
) => {
```

And the logic for handling the abort, also in `handleAsyncRequest`
```
setIsStreaming(true);
for await (const event of readNDJSONStream(responseBody)) {
	if (signal.aborted) {
		// Abort the stream if requested
		break;
	}
```

Create a function for when the button is clicked, and call from where you want, I have mine in the question container, so that the stop button appears instead of the send button, if the answer is streaming

```
const onStopClick = async () => {
	try {
		if (abortController) {
			abortController.abort();
		}
	} catch (e) {
		console.log("An error occurred trying to stop the stream: ", error);
	}
};
```

**Comment by TaylorN15** [comment]
Nice, does yours actually stop the stream on the backend? It seems that most of the time, mine doesn’t. It doesn’t bother me much, but it would be nice to reduce the token usage :)

**Comment by TrentinMarc** [comment]
Hello, sorry for the delay.

Yes, it stops the stream on the backend. Depending on when the chat is stopped, I may save a few tokens.

![image](https://github.com/Azure-Samples/azure-search-openai-demo/assets/48179058/c2324c90-26ba-459f-ae8e-254025520d15)

**Comment by TaylorN15** [comment]
Did you add any code to the backend to handle the abort?

**Comment by TrentinMarc** [comment]
No, I didn't add anything to the backend, as fetch api natively handles AbortController signals.

**Comment by ssjdan27** [comment]
@TrentinMarc I am curious about how you implemented this. I tried following the code layed out here, but when I cancel the stream in the chat I get the following error message on the screen:
`AbortError: BodyStreamBuffer was aborted`

I was wondering if there is a more graceful approach to handling this that you may be aware of.

**Comment by TrentinMarc** [comment]
@ssjdan27 I did not check the recent updates on this project as I stopped working with, but when we were working on it, the simpliest way was to add the signal parameter to the fecth() function performed by chapApi().

Perhaps the asynchronus process has been edited in the meantime ?

If you run out of solution, let's share more informations about your problem :)

**Comment by cforce** [comment]
Please create an Pull Request

**Comment by ssjdan27** [comment]
@cforce I will work on it! I will do my best to get that open within the next few days.

**Comment by ssjdan27** [comment]
@cforce I just opened a PR [here](https://github.com/Azure-Samples/azure-search-openai-demo/pull/1884). Feel free to check it out!
